### DISCUSSION TITLE
How to transform position and rotation from Unreal's coordinate system to Bevy's coordinate system?

### DISCUSSION THREAD

**Discussion by junglie85** [bevyengine/bevy#8311] [Q&A]
I have some the position and rotation of a camera and actor from Unreal and want to display it in Bevy and I cannot change the input data. Unreal has a left-handed coordinate system with `+z` pointing up, `+y` pointing right and `+x` pointing into the screen. Bevy is right-handed, with `+y` pointing up, `+x` pointing right and `-z` pointing into the screen.

Here's an example of the input data:

```
state:
    camera: 
        location:
            x: 2990.3042,
            y: 4194.297,
            z: 1634.4794,
        rotation:
            pitch: 0.0,
            yaw: 0.0,
            roll: -21.098898,
    player:
        location:
            x: 3450.0,
            y: 4020.0,
            z: 1603.4795,
        rotation:
            pitch: 0.0,
            yaw: -0.0,
            roll: -39.9989,
```

Rotation is an Euler rotation, and my assumption is that `pitch` rotates around `y`, `yaw` around `z` and `roll` around `x`.

How can I either (a) convert the position and rotation data so that it can be used in Bevy to display some object in the correct position and orientation relative to the camera, or (b) change Bevy's coordinate system so that it matches that used by Unreal?

My preference would be option (a) and I'd ideally like to be able to do something like:

```
*transform = transform
    .with_translation(Vec3::new(_what_))
    .with_rotation(Quat::from_euler(_what_));
```

When creating the `EulerRot`, in what order would the rotations need to be applied?

**Comment by nicopap** [comment]
## Coordinate translation

Honestly, I'm still myself struggling a bit with this, but I think the following might help.

![Coordinate systems](https://bevy-cheatbook.github.io/img/handedness.png)

Find in the previous table the cell with the
**expected** coordinate, and the cell with the **actual** coordinates.
Then write down the "up", "forward" and "right" axis of the **actual**
coordinates in a column on paper.
Then write down the **expected** coordinates in a column to the right.
If you expected left-handed Y up, and got right-handed Z up,
it should look like this now:
```center
  top     Z  →  Y
  forward Y  →  Z
  right   X  →  X
```

This means you need to go from (x,y,z) to (x,z,y),
which means that your need to swap your Z and Y axis.

### With matrices

You can do it with vector swizzles, but you may also want a transform.
I'm not going to give rotation keys here, because I'm as clueless as you. 
But your software's API let you probably create a transform based on a matrix.

A matrix is an operation on vectors, to go from one 3D vector to another 3D
vector, you need a 3x3 matrix.

The kind of matrix we need is a _rotation_ matrix, that rotates at exactly 90
degree angles.

You can think of a matrix as a table,
where the columns are the "from position" and the rows the "to position".

```
   fx fy fz
tx  1, 0, 0
ty  0, 1, 0
tz  0, 0, 1
```


So for our right-handed Z up to left-handed Y up, conversion from earlier, we
would need the following matrix:

```
    1  0  0
    0  0  1
    0  1  0
```

Check the [`Transform::from_matrix`](https://docs.rs/bevy/latest/bevy/transform/components/struct.Transform.html#method.from_matrix) method for that

## Rotation

You are right regarding pitch/yaw/roll (I think) it should be a relatively straightforward translation into euler angles, though no idea about which order, I always assume XYZ but it might be wrong in your case.
